File: problems/3557-find-maximum-number-of-non-intersecting-substrings/solution.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    public int maxSubstrings(String word) {
        int n = word.length();
        int[] first = new int[26];
        int[] last = new int[26];
        for (int i = 0; i < 26; i++) {
            first[i] = n;
            last[i] = -1;
        }
        for (int i = 0; i < n; i++) {
            int c = word.charAt(i) - 'a';
            first[c] = Math.min(first[c], i);
            last[c] = Math.max(last[c], i);
        }

        List<int[]> intervals = new ArrayList<>();
        for (int i = 0; i < 26; i++) {
            if (first[i] < n) {
                int left = first[i];
                int right = last[i];
                int j = left;
                while (j <= right) {
                    int c = word.charAt(j) - 'a';
                    left = Math.min(left, first[c]);
                    right = Math.max(right, last[c]);
                    j++;
                }
                if (right - left + 1 >= 4) {
                    intervals.add(new int[]{left, right});
                }
            }
        }

        Collections.sort(intervals, (a, b) -> a[1] - b[1]);

        int count = 0;
        int end = -1;
        for (int[] interval : intervals) {
            if (interval[0] > end) {
                count++;
                end = interval[1];
            }
        }

        return count;
    }
}

File: problems/3557-find-maximum-number-of-non-intersecting-substrings/solution.py
class Solution:
    def maxSubstrings(self, word: str) -> int:
        first = {}
        last = {}
        for i, c in enumerate(word):
            if c not in first:
                first[c] = i
            last[c] = i

        intervals = []
        for c in first:
            l, r = first[c], last[c]
            new_l = l
            i = l
            while i <= r:
                char = word[i]
                if first[char] < l:
                    new_l = first[char]
                    l = new_l
                    i = l
                    continue
                r = max(r, last[char])
                i += 1
            intervals.append((l, r))

        intervals.sort(key=lambda x: x[1])
        res = 0
        end = -1
        for l, r in intervals:
            if r - l + 1 >= 4 and l > end:
                res += 1
                end = r
        return res